sherpa/astro/tests/test_astro_data2.py

```diff
@@ -472,20 +472,22 @@ def test_416_a():
     mask = [False, False, False, True, True, True, False, False, False, False]
     assert pha.mask == pytest.approx(mask)
 
+    # With tabStops not set we use ~mask (as this is in channel space).
+    #
     pha.group_counts(3)
 
     # We have a simplified mask
-    mask = [True, True]
+    mask = [False] * 3 + [True] * 2 + [False] * 4
     assert pha.mask == pytest.approx(mask)
 
     # the "full" mask can be retrieved with get_mask
-    mask = [True] * 10
+    mask = [False] * 3 + [True] * 3 + [False] * 4
     assert pha.get_mask() == pytest.approx(mask)
 
-    grouping = [1, -1, -1, -1, -1,  1, -1, -1, -1, -1.]
+    grouping = [0, 0, 0, 1, -1, 1, 0, 0, 0, 0]
     assert pha.grouping == pytest.approx(grouping)
 
-    quality = [0, 0, 0, 0, 0, 2, 2, 2, 2, 2]
+    quality = [0, 0, 0, 0, 0, 2, 0, 0, 0, 0]
```

Comment:
Wait, why has the quality changed? Why is the one in the middle quality 2? I don't see any quality array set anywhere, so I'm confused.

Reply:
So, the group call sets both the grouping and quality arrays. I've added text to the test to explain this, but we can view quality here as

  [3] * 0 + [0, 0, 2] + [0] * 4

The first 3 and last 4 channels are not grouped (since we've excluded them by the filter) and so are set to 0. The only channels we care about are the fourth to sixth channels, which have counts of [2, 1, 1]. The grouping groups them into 2 bins, so we have grouping of [1, -1, 1] to indicate this. The first group (of 2, 1) meets the group criteria (of >= 3) and so has a quality of 0. The last channel has a count of 1, which is < 3, and so has a quality of 3.

When we didn't include the tabstops we carried the 2's to the end of the array because these channels had counts of 0 so the group never hit the >= 3 criteria.